File: elixir/lib/easy/bottle_song.ex
defmodule BottleSong do
  @moduledoc """
  Handles lyrics of the popular children song: Ten Green Bottles
  """

  @numbers %{
    10 => "Ten",
    9 => "Nine",
    8 => "Eight",
    7 => "Seven",
    6 => "Six",
    5 => "Five",
    4 => "Four",
    3 => "Three",
    2 => "Two",
    1 => "One",
    0 => "no"
  }

  @spec recite(pos_integer, pos_integer) :: String.t()
  def recite(start_bottle, take_down) do
    start_bottle..(start_bottle - take_down + 1)
    |> Enum.map(&verse/1)
    |> Enum.join("\n\n")
  end

  defp verse(n) do
    template({@numbers[n], String.downcase(@numbers[n - 1])})
  end

  defp template({left, left_minus_one}) do
    bottles_left_form = if left == "One", do: "bottle", else: "bottles"
    bottles_left_minus_one_form = if left_minus_one == "one", do: "bottle", else: "bottles"

    """
    #{left} green #{bottles_left_form} hanging on the wall,
    #{left} green #{bottles_left_form} hanging on the wall,
    And if one green bottle should accidentally fall,
    There'll be #{left_minus_one} green #{bottles_left_minus_one_form} hanging on the wall.\
    """
  end
end

File: rust/src/easy/bottle_song.rs
pub fn recite(start_bottles: u32, take_down: u32) -> String {
    (1..=start_bottles)
        .rev()
        .take(take_down as usize)
        .map(|i| make_verse(i))
        .collect::<Vec<_>>()
        .join("\n")
}

fn make_verse(i: u32) -> String {
    let amount = index_to_literal(i);
    let amount_minus_one = index_to_literal(i - 1).to_lowercase();

    format!(
        "
{amount} hanging on the wall,
{amount} hanging on the wall,
And if one green bottle should accidentally fall,
There'll be {amount_minus_one} hanging on the wall."
    )
}

fn index_to_literal(i: u32) -> &'static str {
    match i {
        0 => "no green bottles",
        1 => "One green bottle",
        2 => "Two green bottles",
        3 => "Three green bottles",
        4 => "Four green bottles",
        5 => "Five green bottles",
        6 => "Six green bottles",
        7 => "Seven green bottles",
        8 => "Eight green bottles",
        9 => "Nine green bottles",
        10 => "Ten green bottles",
        _ => panic!(),
    }
}
